tools/codegen/api/unboxing.py

```diff
@@ -0,0 +1,199 @@
+#           if (names_opt.has_value()) {
+#                         ~~~~~~~~~~~ <-- Unwrapping optional shell
+#               const c10::IValue names_opt_in = names_opt.value();
+#               const c10::List<c10::IValue> names_list_in = names_opt_in.toList();
+#
+#               std::vector<at::Dimname> names_vec;
+#               for (c10::IValue names_elem: names_list_in) {
+#                                ~~~~~~~~~~~~~~~~~~~~~~~~~ <-- Unrolling list, then convert elements one by one.
+#                   at::Dimname names_base = names_elem.to<at::Dimname>();
+#                   names_vec.push_back(names_base);
+#               }
+#               at::ArrayRef<at::Dimname> names_list_out(names_vec);
+#
+#               names_opt_out = c10::optional<at::ArrayRef<at::Dimname>>(names_list_out);
+#           } else {
+#               names_opt_out = c10::optional<at::ArrayRef<at::Dimname>>();
+#           }
+#       ```
+#    - ScalarType? dtype (similarly for the rest of the arguments)
+#       ```cpp
+#           c10::optional<c10::IValue> dtype_opt = (std::move(peek(stack, 2, 7))).toOptional<c10::IValue>();
+#           c10::optional<at::ScalarType> dtype_opt_out;
+#           if (dtype_opt.has_value()) {
+#               const c10::IValue dtype_opt_in = dtype_opt.value();
+#               at::ScalarType dtype_base = dtype_opt_in.to<at::ScalarType>();
+#                                                        ~~~~~~~~~~~~~~~~~~~~ <-- For base types, convert ivalue to it
+#                                                                                 directly using ".to<T>()" API.
+#               dtype_opt_out = c10::optional<at::ScalarType>(dtype_base);
+#           } else {
+#               dtype_opt_out = c10::optional<at::ScalarType>();
+#           }
+#       ```
+#
+# - Unboxed Kernel Call
+#   ```cpp
+#       auto result_ = torch::empty(
+#           size_list_out,
+#           names_opt_out,
+#           options,
+#           memory_format_opt_out
+#       );
+#   ```
+#
+# - Push Result Back to Stack
+#   ```cpp
+#       drop(stack, 7);
+#       pack(stack, std::move(result_));
+#   ```
+connector = "\n\t"
+
+
+# Convert all the arguments in a NativeFunction to C++ code
+def convert_arguments(args: List[Argument], *, has_tensor_options: bool) -> Tuple[List[Binding], List[str]]:
+    code_list = [f"c10::IValue {args[i].name} = std::move(peek(stack, {i}, {len(args)}));" for i in
+                 range(len(args))] + [""]
+    binding_list = []
+    for i, arg in enumerate(args):
+        unboxed_name, _, code = argumenttype_ivalue_convert(arg.type, arg.name, mutable=arg.is_write)
+        code_list.extend(code)
+        # we need the 'self' argument so method needs to be False, also it doesn't matter if signature faithful or not
+        arg_binding = argument(arg, cpp_no_default_args=set(), method=False, faithful=False,
+                               has_tensor_options=has_tensor_options)
```

Comment:
Okay, so we don't really care about half of the inputs to this function. That's a sign to me that we should be using higher-level api's to do what we're trying to do.

you could directly use `cpp.arguments()` instead, but that also deals directly with `has_tensor_options` and `cpp_no_default_args` stuff.

You can see that the `CppSignature` class abstracts some of that away though, here: https://github.com/pytorch/pytorch/blob/72c972e1e1b4ad838de604e35269e200a70db5f2/tools/codegen/api/types.py#L422. You can pass in the NativeFunction and nothing else, and you'll get a `CppSignature` with all of the correct "default" properties for that NativeFunction (including the fact that say, `cpp_no_default_args` was generated from `f.cpp_no_default_args`.

So when I see that, it tells me that maybe this `convert_arguments()` function should just directly take in the entire `NativeFunction`. Then you can do something like:
```
def convert_arguments(f: NativeFunction) -> Tuple[List[Binding], List[str]]:
    cpp_args = CppSignature.from_native_function(f).most_faithful_signature().arguments()
    binding_list = []
    code_list = []
    for i, arg in enumerate(cpp_args):
        unboxed_name, code = argumenttype_ivalue_convert(arg.argument.type, arg.argument.name, mutable=arg.argument.is_write)
        # generate code for unboxing each ivalue
        code_list.append(f"""\n
c10::IValue {cpp_args[i].name} = std::move(peek(stack, {i}, {len(cpp_args)}));
{code}"""
      
        binding_list.append(arg.with_name(unboxed_name))
```

Reply:
Or alternatively, you could probably directly pass the `CppSignature` instead of the `NativeFunction` since it carries all of the same information, and you need to re-use the `CppSignature` in `gen_unboxing` later on

Reply:
I'll need to pass a `NativeFunction` since I need to set `method=False` for the `CppSignature`. This makes sure the self argument is present.